sysconfig/cloudinit.go

```diff
@@ -76,21 +80,59 @@ func installCloudInitCfg(src, targetdir string) error {
 	return nil
 }
 
-// TODO:UC20: - allow cloud.conf coming from the gadget
-//            - think about if/what cloud-init means on "secured" models
+// installGadgetCloudInitCfg installs a single cloud-init config file from the
+// gadget snap to the /etc/cloud config dir as "80_device_gadget.cfg".
+func installGadgetCloudInitCfg(src, targetdir string) error {
+	ubuntuDataCloudCfgDir := filepath.Join(ubuntuDataCloudDir(targetdir), "cloud.cfg.d/")
+	if err := os.MkdirAll(ubuntuDataCloudCfgDir, 0755); err != nil {
+		return fmt.Errorf("cannot make cloud config dir: %v", err)
+	}
+
+	configFile := filepath.Join(ubuntuDataCloudCfgDir, "80_device_gadget.cfg")
+	return osutil.CopyFile(src, configFile, 0)
+}
+
 func configureCloudInit(opts *Options) (err error) {
 	if opts.TargetRootDir == "" {
 		return fmt.Errorf("unable to configure cloud-init, missing target dir")
 	}
 
-	switch opts.CloudInitSrcDir {
-	case "":
-		// disable cloud-init by default using the writable dir
-		err = DisableCloudInit(WritableDefaultsDir(opts.TargetRootDir))
-	default:
-		err = installCloudInitCfg(opts.CloudInitSrcDir, WritableDefaultsDir(opts.TargetRootDir))
+	// first check if cloud-init should be disallowed entirely
+	if !opts.AllowCloudInit {
+		if err := DisableCloudInit(WritableDefaultsDir(opts.TargetRootDir)); err != nil {
+			return err
+		}
+	}
+
+	// next check if there is a gadget cloud.conf to install
+	gadgetCloudConf := filepath.Join(opts.GadgetDir, "cloud.conf")
+	if osutil.FileExists(gadgetCloudConf) {
+		// then copy / install the gadget config and return without considering
+		// CloudInitSrcDir
+		if err := installGadgetCloudInitCfg(gadgetCloudConf, WritableDefaultsDir(opts.TargetRootDir)); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// TODO:UC20: implement filtering of files from src when specified via a
+	//            specific Options for i.e. signed grade and MAAS, etc.
+
+	// finally check if there is a cloud-init src dir we should copy config
+	// files from
+
+	if opts.CloudInitSrcDir != "" {
+		if err := installCloudInitCfgDir(opts.CloudInitSrcDir, WritableDefaultsDir(opts.TargetRootDir)); err != nil {
```

Comment:
Should we return here too (just like we did in the previous "if"s)? I know it has no consequences currently just wondering as it would make things more symmetric.

Reply:
sure, I did this now